### PR TITLE
Support mysql prepare statement command

### PR DESCRIPTION
**Pull request by adriansr**
This is a cleanup of #5663 (Support mysql prepare statement command by @ggg-geyi)

### CODE REVIEW

**Comment by ggg-geyi** [comment]
I have checked these commits. Thanks  for your help.

File: packetbeat/protos/mysql/mysql.go [hunk 1]
```diff
@@ -675,8 +805,206 @@ func (mysql *mysqlPlugin) receivedMysqlResponse(msg *mysqlMessage) {
...
+		//  FIELD_TYPE_LONGLONG
+		case 0x08:
+			if dataLen < paramOffset+8 {
+				logp.Debug("mysql", "Data too small")
+				return nil
+			}
+			valueString := strconv.FormatInt(int64(data[paramOffset])|int64(data[paramOffset+1])<<8|
+				int64(data[paramOffset+2])<<16|int64(data[paramOffset+3])<<24|
+				int64(data[paramOffset+4])<<32|int64(data[paramOffset+5])<<40|
+				int64(data[paramOffset+6])<<48|int64(data[paramOffset+7])<<56, 10)
```
Comment (cwurm) [inline packetbeat/protos/mysql/mysql.go]: Should be `binary.littleEndian.Uint64`?

File: packetbeat/protos/mysql/mysql.go [hunk 2]
```diff
@@ -675,8 +805,206 @@ func (mysql *mysqlPlugin) receivedMysqlResponse(msg *mysqlMessage) {
...
+			paramOffset += paramLen
+		// FIELD_TYPE_VAR_STRING
+		// FIELD_TYPE_BLOB
+		// FIELD_TYPE_STRING
+		case 0xf6, 0xfc, 0xfd, 0xfe:
+			paramLen := int(data[paramOffset])
+			paramOffset++
+			switch paramLen {
+			case 0xfc: /* 252 - 64k chars */
+				paramLen16 := int(data[paramOffset]) | int(data[paramOffset+1])<<8
```
Comment (cwurm) [inline packetbeat/protos/mysql/mysql.go]: Should be `binary.littleEndian.Uint16`?

File: packetbeat/protos/mysql/mysql.go [hunk 3]
```diff
@@ -611,9 +679,46 @@ func (mysql *mysqlPlugin) receivedMysqlRequest(msg *mysqlMessage) {
...
+				trans.query = "Request Execute Statement"
+				trans.notes = append(trans.notes, "The actual query being used is unknown")
+				trans.requestRaw = msg.query
+				trans.bytesIn = msg.size
+				return
+			}
+		} else if msg.typ == mysqlCmdStmtClose {
+			delete(stmts, trans.statementID)
+			trans.query = "CmdStmtClose"
+			//mysql.transactions.Delete(tuple.Hashable())
```
[the remark this answers is not in the document]
Comment (ggg-geyi) [reply packetbeat/protos/mysql/mysql.go]: Right，we‘re not going to receive a response.
And if we get the next cmd and we will get the message as `Two requests without a Response. Dropping old request`.
We should delete the transaction on close.

**Comment by adriansr** [comment]
@ggg-geyi I am currently working in some problems I found with the mysql parser (not caused by your code), I will fix them and then take care of the changes requested in this PR and merging it.

**Comment by andrewkroh** [comment]
I believe this addresses both of these issues and they can be closed by this PR.

- https://github.com/elastic/beats/issues/5608
- https://github.com/elastic/beats/issues/683

**Comment by adriansr** [comment]
@cwurm I addressed all your comments, do you mind to give it another look?

File: packetbeat/tests/system/test_0064_mysql_prepare_statement.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from packetbeat import BaseTest
+
+
+class Test(BaseTest):
+
+    def test_mysql_prepare_statement(self):
+        self.render_config_template(
+            mysql_ports=[3307]
+        )
+        self.run_packetbeat(pcap="mysql_prepare_statement.pcap",
```
Comment (andrewkroh) [inline packetbeat/tests/system/test_0064_mysql_prepare_statement.py]: 👍

File: packetbeat/protos/mysql/mysql.go [hunk 5]
```diff
@@ -140,6 +152,10 @@ type mysqlPlugin struct {
 	transactions       *common.Cache
 	transactionTimeout time.Duration
 
+	// prepare statements cache
+	prepareStatements       *common.Cache
+	prepareStatementTimeout time.Duration
```
Comment (urso) [inline packetbeat/protos/mysql/mysql.go]: Prepared statements are accessed via address tuple. A mysql connection is allowed to only have this many prepared statements (applications requiring more or dynamic set of prepared statements tend to implement a Cache of active statements).

Why store the statements globally and time them out globally? I wonder if we want to move the statements to the TCP connection state.
Advantage is:  if the connection is gone all known prepared statements are gone as well
Disadvantage: if connection is dropped (internal timeout or parse error after gaps), then we loose the current state.

Alternatively bind to connections + timeout only if connection is dropped/errored:
- If TCP connection is properly closed we can can drop the state
- If connection is dropped we put the connection it's statements on timeout
- If new connection without TCP handshake is found we check if set of prepared statements is available for address pair and assign it to the current connection (stopping timeout).

**Comment by cwurm** [comment]
> @cwurm I addressed all your comments, do you mind to give it another look?

LGTM. Sorry I took so long to get to this.

See @urso's comment though - sounds like a good idea to me, but not sure if it's worth putting into this already very long running PR or if it should rather be handled as a follow-up. Anything is good with me.

**Comment by urso** [comment]
> See @urso's comment though - sounds like a good idea to me, but not sure if it's worth putting into this already very long running PR or if it should rather be handled as a follow-up. Anything is good with me.

This might also need some improvements in signaling a plugin about connection state. +1 for changes/improvements to come later. Let's open an issue, though.

**Comment by adriansr** [comment]
Thanks @urso I will follow up with a new PR.

**Pull request merged**